pyogrio/_err.pyx

```diff
@@ -132,53 +144,88 @@ exception_map = {
     13: CPLE_AWSObjectNotFoundError,
     14: CPLE_AWSAccessDeniedError,
     15: CPLE_AWSInvalidCredentialsError,
-    16: CPLE_AWSSignatureDoesNotMatchError
+    16: CPLE_AWSSignatureDoesNotMatchError,
+    17: CPLE_AWSError
 }
 
 
-cdef inline object exc_check():
-    """Checks GDAL error stack for fatal or non-fatal errors
+cdef inline object check_last_error():
+    """Checks if the last GDAL error was a fatal or non-fatal error.
+
+    When a non-fatal error is found, an appropriate exception is raised.
+
+    When a fatal error is found, SystemExit is called.
+
     Returns
     -------
     An Exception, SystemExit, or None
     """
-    cdef const char *msg_c = NULL
-
     err_type = CPLGetLastErrorType()
     err_no = CPLGetLastErrorNo()
-    err_msg = CPLGetLastErrorMsg()
+    err_msg = get_last_error_msg()
+
+    if err_type == CE_Failure:
+        CPLErrorReset()
+        return exception_map.get(
+            err_no, CPLE_BaseError)(err_type, err_no, err_msg)
+
+    if err_type == CE_Fatal:
+        return SystemExit("Fatal error: {0}".format((err_type, err_no, err_msg)))
 
-    if err_msg == NULL:
-        msg = "No error message."
     else:
-        # Reformat messages.
-        msg_b = err_msg
+        return
+
+
+cdef get_last_error_msg():
+    """Checks GDAL error stack for the latest error message.
+
+    Returns
+    -------
+    An error message or empty string
+
+    """
+    return clean_error_message(CPLGetLastErrorMsg())
+
+
+cdef clean_error_message(const char* err_msg):
+    """Cleans up error messages from GDAL.
 
+    Parameters
+    ----------
+    err_msg : const char*
+        The error message to clean up.
+
+    Returns
+    -------
+    str
+        The cleaned up error message or empty string
+    """
+    if err_msg != NULL:
+        # Reformat message.
+        msg_b = err_msg
         try:
-            msg = msg_b.decode('utf-8')
+            msg = msg_b.decode("utf-8")
             msg = msg.replace("`", "'")
             msg = msg.replace("\n", " ")
         except UnicodeDecodeError as exc:
-            msg = f"Could not decode error message to UTF-8.  Raw error: {msg_b}"
+            msg = f"Could not decode error message to UTF-8. Raw error: {msg_b}"
 
-    if err_type == 3:
-        CPLErrorReset()
-        return exception_map.get(
-            err_no, CPLE_BaseError)(err_type, err_no, msg)
+    else:
+        msg = ""
```

Comment:
Do we want to keep "No error message." here? Or we don't want that for the case that we are stacking multiple errors where not every error might have a message?

Reply:
That's why I retained the empty string. I'm not sure if having no error message actually happens... it seems there is no test for it either...

Reply:
I moved the check so "No error message." will be shown in the typical cases where there would be an exception without message...